File: cp1/rpi_pubsub.cpp
//run on raspberry pi (c++版本) 可work
#include "ros/ros.h" 
#include "std_msgs/Int32.h" 
#include <iostream> 
#include <stdio.h>
      
int flag = 1; 

void number_callback(const std_msgs::Int32::ConstPtr& msg) {
    printf("message from Arduino is %d\n", msg->data); 
    flag = 1;
}

int main(int argc, char **argv) 
{ 
 
    ros::init(argc, argv, "rpi_node");

    ros::NodeHandle node_obj;

    ros::AsyncSpinner spinner(0);
    spinner.start();
    
    ros::Publisher rpi_publisher = node_obj.advertise<std_msgs::Int32>("topic_rpisend", 10); 
    ros::Subscriber rpi_subscriber = node_obj.subscribe("topic_arduinosend", 10, number_callback);
    ros::Rate loop_rate(1); 

    
    while (ros::ok()) {
        std_msgs::Int32 msg;
        
        if (flag == 1){
            std::cout << "user's input is :";
            std::cin >> msg.data;
            rpi_publisher.publish(msg);
            flag = 0;
        }
               
        
        ros::spinOnce(); 
        loop_rate.sleep();

    } 
    return 0;
}

File: cp1/rpi_pubsub.py
#!/usr/bin/env python
# run on raspberry pi (python版) 可work

import rospy
 
from std_msgs.msg import Int32

#global rec_flag
 
return_msg=None
 
#define the display text
def number_callback(msg):
    global rec_flag
    print "message from Arduino is : {}" .format(msg) 
    rec_flag = True
 
 
if __name__=='__main__':
    rospy.init_node('rpi_node')
    rospy.Subscriber("topic_arduinosend",Int32, number_callback)
    rpi_publisher = rospy.Publisher("topic_rpisend", Int32, queue_size=10)
    rate=rospy.Rate(1)

    global rec_flag
    rec_flag = True

    while not rospy.is_shutdown():
        if rec_flag == True:
            msg = input("user's input is: ")
            rpi_publisher.publish(msg)
            rec_flag = False
        
     
 
   
